sys/netlink/netlink_snl.h

```diff
@@ -63,7 +63,7 @@
 
 #define	_NLA_END(_start, _len)	((char *)(_start) + (_len))
 #define NLA_FOREACH(_attr, _start, _len)      \
-        for (_attr = (_start);		\
+        for (_attr = (typeof(_attr))(_start);		\
```

Comment:
Does this work with C still?

Reply:
typeof is a compiler extension, it was added in C23.

Reply:
So no then. We don't enable c23 for the kernel at all.

Reply:
changed to use `__typeof` instead

Reply:
Uh how about doing...
```cpp
#ifdef __cplusplus
# define NLA_FOREACH(_attr, _start, _len)      \
        for (_attr = (decltype(_start))(_start);	\
		((char *)_attr < _NLA_END(_start, _len)) && \
		((char *)NLA_NEXT(_attr) <= _NLA_END(_start, _len));	\
		_attr =  NLA_NEXT(_attr))
#else
# define NLA_FOREACH(_attr, _start, _len)      \
        for (_attr = (_start);		\
		((char *)_attr < _NLA_END(_start, _len)) && \
		((char *)NLA_NEXT(_attr) <= _NLA_END(_start, _len));	\
		_attr =  NLA_NEXT(_attr))
#endif
```

This one would be valid C and C++ code (without using any compiler extensions)... but the only issue (visually) is that two similar macros, one for C and one for C++

Reply:
Splitting it into two macros might make future maintenance inconvenience.

How about I change it to `(struct nlattr *)` just like `NLA_NEXT` at line 62?
`NLA_FOREACH` macro uses `NLA_NEXT` internally, and `NLA_NEXT` is directly cast as `(struct nlattr *)`.

Reply:
> So no then. We don't enable c23 for the kernel at all.

`typeof()` should be fine in practice. Compiling the following fragment on compiler explorer says clang >= 11 and gcc >= 10 support it with default flags.

```C
int a;
typeof(a) b;
```